core/arch/arm/kernel/thread_spmc.c

```diff
@@ -326,46 +327,103 @@ static bool is_my_uuid(uint32_t w0, uint32_t w1, uint32_t w2, uint32_t w3)
 {
-	uint32_t ret_fid = 0;
+	fpi->id = endpoint_id;
+	fpi->execution_context = CFG_TEE_CORE_NB_CORE;
+	fpi->partition_properties = BIT(0) | BIT(1);
+}
+
+static uint32_t handle_partition_info_get_all(int *rc, struct ffa_rxtx *rxtx)
+{
+	struct ffa_partition_info *fpi = NULL;
+
+	if (!rxtx->size || !rxtx->tx_is_mine) {
+		if (rxtx->size)
+			*rc = FFA_BUSY;
+		else
+			*rc = FFA_DENIED; /* TX buffer not setup yet */
+		return FFA_ERROR;
+	}
+
+	fpi = rxtx->tx;
+
+	/* Add OP-TEE SP */
+	spmc_create_partition_entry(fpi, my_endpoint_id);
+	rxtx->tx_is_mine = false;
+	*rc = 1;
+	fpi++;
+
+	if (IS_ENABLED(CFG_SECURE_PARTITION)) {
+		size_t max_elements = (rxtx->size / sizeof(*fpi)) - 1;
+
+		*rc += sp_get_all_partitions(fpi, max_elements);
+	}
+	return FFA_SUCCESS_32;
+}
+
+void spmc_handle_partition_info_get(struct thread_smc_args *args,
+				    struct ffa_rxtx *rxtx)
+{
+	uint32_t ret_fid = FFA_ERROR;
 	int rc = 0;
+	uint32_t endpoint_id = my_endpoint_id;
 
-	if (!is_nil_uuid(args->a1, args->a2, args->a3, args->a4) &&
-	    !is_my_uuid(args->a1, args->a2, args->a3, args->a4)) {
-		ret_fid = FFA_ERROR;
-		rc = FFA_INVALID_PARAMETERS;
+	cpu_spin_lock(&rxtx->spinlock);
+	if (is_nil_uuid(args->a1, args->a2, args->a3, args->a4)) {
+		ret_fid = handle_partition_info_get_all(&rc, rxtx);
 		goto out;
 	}
 
-	cpu_spin_lock(&rxtx->spinlock);
+	if (!is_my_uuid(args->a1, args->a2, args->a3, args->a4)) {
+		if (IS_ENABLED(CFG_SECURE_PARTITION)) {
+			uint32_t uuid_array[4] = { 0 };
+			TEE_UUID uuid = { };
+			TEE_Result res = TEE_SUCCESS;
+
+			/*
+			 * Incoming UUID is little-endian as per
+			 * SMCCC v1.2 Sect. 5.3.
+			 */
+			uuid_array[0] = (uint32_t)args->a1;
+			uuid_array[1] = (uint32_t)args->a2;
+			uuid_array[2] = (uint32_t)args->a3;
+			uuid_array[3] = (uint32_t)args->a4;
+			tee_uuid_from_octets(&uuid, (uint8_t *)uuid_array);
+
+			res = sp_get_session_id(&uuid, &endpoint_id);
+			if (res != TEE_SUCCESS) {
+				ret_fid = FFA_ERROR;
+				rc = FFA_INVALID_PARAMETERS;
+				goto out;
+			}
+		} else {
+			ret_fid = FFA_ERROR;
+			rc = FFA_INVALID_PARAMETERS;
+			goto out;
+		}
+	}
+
 	if (rxtx->size && rxtx->tx_is_mine) {
 		struct ffa_partition_info *fpi = rxtx->tx;
 
-		fpi->id = my_endpoint_id;
-		fpi->execution_context = CFG_TEE_CORE_NB_CORE;
-		/*
-		 * Supports receipt of direct requests.
-		 * Can send direct requests.
-		 */
-		fpi->partition_properties = BIT(0) | BIT(1);
-
+		spmc_create_partition_entry(fpi, endpoint_id);
 		ret_fid = FFA_SUCCESS_32;
-		rc = 1;
 		rxtx->tx_is_mine = false;
+		rc = 1;
 	} else {
 		ret_fid = FFA_ERROR;
 		if (rxtx->size)
 			rc = FFA_BUSY;
 		else
 			rc = FFA_DENIED; /* TX buffer not setup yet */
 	}
-	cpu_spin_unlock(&rxtx->spinlock);
 
 out:
-	spmc_set_args(args, ret_fid, FFA_PARAM_MBZ, rc, FFA_PARAM_MBZ,
-		      FFA_PARAM_MBZ, FFA_PARAM_MBZ);
+	cpu_spin_unlock(&rxtx->spinlock);
+	spmc_set_args(args, ret_fid, FFA_PARAM_MBZ, rc,
+		      FFA_PARAM_MBZ, FFA_PARAM_MBZ, FFA_PARAM_MBZ);
```

Comment:
change not needed for the 2 last line.

Reply:
I only added cpu_spin_unlock(). I do think this is needed to make sure that we always release the lock

Reply:
you also moved an occurrence of `FFA_PARAM_MBZ`.